src/app/icd/client/ICDRefreshKeyInfo.h

```diff
@@ -0,0 +1,105 @@
+#include <lib/support/CodeUtils.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+typedef Crypto::SensitiveDataBuffer<Crypto::kAES_CCM128_Key_Length> RefreshKeyBuffer;
+
+class CheckInDelegate;
+/**
+ * @brief ICDRefreshKeyInfo contains all the data and methods needed for key refresh and re-registration of an ICD client.
+ */
+class ICDRefreshKeyInfo : public CommandSender::Callback
+{
+public:
+    ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, const ICDClientInfo aICDClientInfo,
+                      const ICDClientStorage * aICDClientStorage);
+
+    // CommandSender callbacks
+    void OnResponse(chip::app::CommandSender * apCommandSender, const chip::app::ConcreteCommandPath & aPath,
+                    const chip::app::StatusIB & aStatus, chip::TLV::TLVReader * aData) override;
+    void OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError) override;
+    void OnDone(chip::app::CommandSender * apCommandSender) override;
+
+    // CASE session callbacks
+    /**
+     * @brief Callback received on successfully establishing a CASE session in order to re-register the client with the peer node
+     * using a new key to avoid counter rollover problems.
+     *
+     * @param[in] context context of the client establishing the CASE session
+     * @param[in] exchangeMgr exchange manager to use for the re-registration
+     * @param[in] sessionHandle session handle to use for the re-registration
+     */
+    static void HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                      const SessionHandle & sessionHandle);
+    /**
+     * @brief Callback received on failure to establish a CASE session in order to re-register the client with the peer node using a
+     * new key to avoid counter rollover problems.
+     *
+     * @param[in] context context of the client establishing the CASE session
+     * @param[in] peerId Scoped Node ID of the peer node
+     * @param[in] err failure reason
+     */
+    static void HandleDeviceConnectionFailure(void * context, const ScopedNodeId & peerId, CHIP_ERROR err);
+
+    /**
+     * @brief Used by the application to set a new key to avoid counter rollover problems.
+     *
+     * @param[in] clientInfo clientInfo object
+     * @param[in] keyData New key data to use to re-register the client with the server
+     * @param[in] exchangeMgr exchange manager to use for the re-registration
+     * @param[in] sessionHandle session handle to use for the re-registration
+     */
+    CHIP_ERROR RegisterClientWithNewKey(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle);
+
+    /**
+     * @brief Sets up a CASE session to the peer for re-registering a client with the peer when a key refresh is required to avoid
+     * ICD counter rollover. Returns error if we did not even manage to kick off a CASE attempt.
+     */
+    void EstablishSessionToPeer();
```

Comment:
Why is this public API?

Reply:
They need not be public. I'll make them private.